### PR TITLE
Support OpenEndedDictionary (Mozmap) in the Headers

### LINKED ISSUE #11897: Implement the Headers API
Specification: https://fetch.spec.whatwg.org/#headers-class
Documentation: http://doc.servo.org/servo/script/dom/index.html#adding-a-new-dom-interface

This is assigned to our RGSoC team, so please talk with me before starting to work on any part of it.

### PR DESCRIPTION
**Pull request by jeenalee**
This PR will support OpenEndedDictionary (based on MozMap implementation from #13332) in the Headers API.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #11897  (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13356)

### CODE REVIEW

File: components/script/dom/request.rs [hunk 1]
```diff
@@ -14,7 +14,7 @@ use dom::bindings::codegen::Bindings::RequestBinding::RequestMethods;
 use dom::bindings::codegen::Bindings::RequestBinding::RequestMode;
 use dom::bindings::codegen::Bindings::RequestBinding::RequestRedirect;
 use dom::bindings::codegen::Bindings::RequestBinding::RequestType;
-use dom::bindings::codegen::UnionTypes::HeadersOrByteStringSequenceSequence;
+use dom::bindings::codegen::UnionTypes::HeadersOrByteStringSequenceSequenceOrByteStringMozMap  as HeadersInitType;
```
Comment (jdm) [inline components/script/dom/request.rs]: nit: extra space before as.

File: components/script/dom/request.rs [hunk 2]
```diff
@@ -796,13 +796,18 @@ impl Into<RequestRedirect> for NetTraitsRequestRedirect {
...
-            HeadersOrByteStringSequenceSequence::Headers(h.clone()),
-        &HeadersOrByteStringSequenceSequence::ByteStringSequenceSequence(ref b) =>
-            HeadersOrByteStringSequenceSequence::ByteStringSequenceSequence(b.clone()),
-        }
+        &HeadersInitType::Headers(ref h) =>
+            HeadersInitType::Headers(h.clone()),
+        &HeadersInitType::ByteStringSequenceSequence(ref b) =>
+            HeadersInitType::ByteStringSequenceSequence(b.clone()),
+        &HeadersInitType::ByteStringMozMap(ref m) =>
+        {
```
Comment (jdm) [inline components/script/dom/request.rs]:
```rust
&HeadersInitType::ByteStringMozMap(ref m) =>
    HeadersInitType::ByteStringMozMap(m.clone())
```

**Review by jdm — CHANGES_REQUESTED** [review]
I'm operating under the assumption that all of the tests now timing out require Promises. These changes look good!

File: components/script/dom/request.rs [hunk 3]
```diff
@@ -14,7 +14,7 @@ use dom::bindings::codegen::Bindings::RequestBinding::RequestMethods;
 use dom::bindings::codegen::Bindings::RequestBinding::RequestMode;
 use dom::bindings::codegen::Bindings::RequestBinding::RequestRedirect;
 use dom::bindings::codegen::Bindings::RequestBinding::RequestType;
-use dom::bindings::codegen::UnionTypes::HeadersOrByteStringSequenceSequence;
+use dom::bindings::codegen::UnionTypes::HeadersOrByteStringSequenceSequenceOrByteStringMozMap as HeadersInitType;
```
Comment (KiChjang) [inline components/script/dom/request.rs]: Hmm... `dom::bindings::codegen::Bindings::RequestBinding` should re-export this union type as `HeadersInit`. Have you tried importing that instead?

**Comment by jdm** [comment]
I'm going to hold off merging this until after Promises, since there's a lot of overlap between the test result changes.

File: components/script/dom/request.rs (hunk 3 above)
Comment (jeenalee) [inline components/script/dom/request.rs]: You're right! Fixing it right now. Thanks!

**Comment by jdm** [comment]
If you rebase, I will merge :)

**Comment by jdm** [comment]
Oh, and go ahead and squash the commits into one.

**Pull request merged**